### PR TITLE
FSAF Module

### PR DESCRIPTION
**Pull request by zccstig**
Reimplementation of FSAF Module

### CODE REVIEW

File: mmdet/models/anchorfree_heads/fsaf_head.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+                pos_boxes[:, 2:] *= cfg.pos_scale
+                pos_boxes = xcycwh2xyxy(pos_boxes)
+                ig_x1 = torch.floor(ig_boxes[:, 0]).clamp(0, w - 1).int()
+                ig_y1 = torch.floor(ig_boxes[:, 1]).clamp(0, h - 1).int()
+                ig_x2 = torch.ceil(ig_boxes[:, 2]).clamp(1, w).int()
+                ig_y2 = torch.ceil(ig_boxes[:, 3]).clamp(1, h).int()
+                pos_x1 = torch.floor(pos_boxes[:, 0]).clamp(0, w - 1).int()
+                pos_y1 = torch.floor(pos_boxes[:, 1]).clamp(0, h - 1).int()
+                pos_x2 = torch.ceil(pos_boxes[:, 2]).clamp(1, w).int()
+                pos_y2 = torch.ceil(pos_boxes[:, 3]).clamp(1, h).int()
```
Comment (hellock) [inline mmdet/models/anchorfree_heads/fsaf_head.py]: There are quite a few repeated code blocks for computing a proportional region, which can be extracted to a method, similar to https://github.com/open-mmlab/mmdetection/pull/594/files#diff-6c0c212b893524716695388f57ab71d4R7.

File: mmdet/models/anchorfree_heads/fsaf_head.py [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    def _init_layers(self):
+        self.relu = nn.ReLU(inplace=True)
+        self.cls_convs = nn.ModuleList()
+        self.reg_convs = nn.ModuleList()
+        for i in range(self.stacked_convs):
+            chn = self.in_channels if i == 0 else self.feat_channels
+            self.cls_convs.append(
+                nn.Conv2d(chn, self.feat_channels, 3, stride=1, padding=1))
+            self.reg_convs.append(
+                nn.Conv2d(chn, self.feat_channels, 3, stride=1, padding=1))
```
[the remark this answers is not in the document]
Comment (zccstig) [reply mmdet/models/anchorfree_heads/fsaf_head.py]: Using ConvModule causes the following runtime error:

RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [4, 256, 10, 7]], which is output 0 of ReluBackward1, is at version 2; expected version 1 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).

**Comment by laycoding** [comment]
the pred of regression outputs might be negative, which may cause the select iou greater than 1.
maybe you could clamp it by
`pred = pred.clamp(min=1e-7)`

**Comment by zccstig** [comment]
@laycoding Regression output is guaranteed to be nonnegative due to the relu function (see line 94 in fsaf_head.py)

**Comment by laycoding** [comment]
@zccstig thanks, I didn't see it

**Comment by Yuesheng321** [comment]
Would you provide the PTH and the testing result on the coco data set ?

**Comment by ZwwWayne** [comment]
Already supported in V2.0 through https://github.com/open-mmlab/mmdetection/pull/2521.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @zccstig ！We are grateful for your efforts in helping improve mmdetection open-source project during your personal time. 

Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）

Thank you again for your contribution❤

**Pull request closed without merging**